public/app/plugins/panel/datagrid/DataGridPanel.tsx

```diff
@@ -0,0 +1,288 @@
+  const onColumnResizeEnd = (column: GridColumn, newSize: number, colIndex: number, newSizeWithGrow: number) => {
+    dispatch({ type: DatagridActionType.columnResizeEnd });
+  };
+
+  const closeContextMenu = () => {
+    dispatch({ type: DatagridActionType.closeContextMenu });
+  };
+
+  const onDeletePressed = (selection: GridSelection) => {
+    if (selection.current && selection.current.range) {
+      publishSnapshot(clearCellsFromRangeSelection(frame, selection.current.range), id);
+      return true;
+    }
+
+    if (selection.rows) {
+      publishSnapshot(deleteRows(frame, selection.rows.toArray()), id);
+      return true;
+    }
+
+    return false;
+  };
+
+  const onCellContextMenu = (cell: Item, event: CellClickedEventArgs) => {
+    event.preventDefault();
+    dispatch({ type: DatagridActionType.openCellContextMenu, payload: { event, cell } });
+  };
+
+  const onHeaderContextMenu = (columnIndex: number, event: HeaderClickedEventArgs) => {
+    event.preventDefault();
+    dispatch({ type: DatagridActionType.openHeaderContextMenu, payload: { event, columnIndex } });
+  };
+
+  const onHeaderMenuClick = (col: number, screenPosition: Rectangle) => {
+    dispatch({
+      type: DatagridActionType.openHeaderDropdownMenu,
+      payload: { screenPosition, columnIndex: col, value: frame.fields[col].name },
+    });
+  };
+
+  const onColumnMove = (from: number, to: number) => {
+    const newFrame = new MutableDataFrame(frame);
+    const field = newFrame.fields[from];
+    newFrame.fields.splice(from, 1);
+    newFrame.fields.splice(to, 0, field);
+
+    dispatch({ type: DatagridActionType.columnMove, payload: { from, to } });
+    publishSnapshot(newFrame, id);
+  };
+
+  const onRowMove = (from: number, to: number) => {
+    const newFrame = new MutableDataFrame(frame);
+
+    for (const field of newFrame.fields) {
+      const values = field.values.toArray();
+      const value = values[from];
+      values.splice(from, 1);
+      values.splice(to, 0, value);
+      field.values = new ArrayVector(values);
+    }
+
+    publishSnapshot(newFrame, id);
+  };
+
+  const onColumnRename = () => {
+    dispatch({ type: DatagridActionType.showColumnRenameInput });
+  };
+
+  const onRenameInputBlur = (columnName: string, columnIdx: number) => {
+    const newFrame = new MutableDataFrame(frame);
+    newFrame.fields[columnIdx].name = columnName;
+
+    dispatch({ type: DatagridActionType.hideColumnRenameInput });
+    publishSnapshot(newFrame, id);
+  };
+
+  const onSearchClose = () => {
+    dispatch({ type: DatagridActionType.closeSearch });
+  };
+
+  const onGridSelectionChange = (selection: GridSelection) => {
+    dispatch({ type: DatagridActionType.multipleCellsSelected, payload: { selection } });
+  };
+
+  if (!frame) {
+    return <PanelDataErrorView panelId={id} fieldConfig={fieldConfig} data={data} />;
+  }
+
```

Comment:
I really struggle with these large components with so many callback functions & state logic. 

I think using the StateManagerBase pattern would have worked really nicely here as all these callbacks ALL the state logic could be moved outsider render path (and also tested easily) and be handled together, not this mix of logic and callbacks here and redux actions spreads it out and still leaves a large react component full with mostly inline functions with state logic. etc no need to create callback functions that then dispatch actions, you can bind the react prop directly to the function. 

See example here  https://github.com/grafana/grafana/blob/842ce144292bdf6b51ba2e13961c0986969005e4/public/app/features/dashboard/components/HelpWizard/SupportSnapshotService.ts#L100

But just my preference, I know other see no problem with a render function with a bunch of state callbacks :)